sparse/client.go

```diff
@@ -139,334 +109,243 @@ func connect(host, port string, timeout int) net.Conn {
-			lrange = <-local
-			rrange = <-remote
-		} else {
-			// Should never happen
-			log.Fatal("processDiff internal error")
-			return
 		}
 	}
-	log.Info("Finished processing file diff")
 
-	status := true
-	err = <-errStream
-	if err != nil {
-		log.Error("Sync client file load aborted:", err)
-		status = false
-	}
-	// make sure we finished consuming dst hashes
-	status = <-netInStreamDone && status // netDstReceiver finished
-	log.Info("Finished consuming remote file hashes, status=", status)
-
-	// Send end of transmission
-	netStream <- diffChunk{true, DataInterval{FileInterval{SparseIgnore, Interval{0, 0}}, make([]byte, 0)}}
-
-	// get network sender status
-	net := <-netStatus
-	log.Info("Finished sending file diff of", net.byteCount, "(bytes), status=", net.status)
-	if !net.status {
-		err = errors.New("netwoek transfer failure")
-		return
+	// special case, the whole file is a hole
+	if extCount == 0 && fileSize != 0 {
+		// report hole
+		holeInterval = Interval{0, fileSize}
+		log.Debugf("The file is a hole: %s", holeInterval)
+
+		// syncing hole interval
+		err := SyncHoleInterval(encoder, decoder, holeInterval)
+		if err != nil {
+			log.Debugf("SyncHoleInterval failed: %s", holeInterval)
+			return err
+		}
 	}
 
-	var statusRemote bool
-	err = decoder.Decode(&statusRemote)
+	encoder.Encode(requestHeader{requestMagic, syncDone})
+	return nil
+}
+
+func sendSyncRequest(encoder *gob.Encoder, decoder *gob.Decoder, path string, size int64) (bool, error) {
+	err := encoder.Encode(requestHeader{requestMagic, syncRequestCode})
 	if err != nil {
-		log.Fatal("Cient protocol remote status error:", err)
-		return
+		log.Debug("send requestHeader failed: ", err)
+		return false, err
 	}
-	if !statusRemote {
-		err = errors.New("failure on remote sync site")
-		return
+	err = encoder.Encode(path)
+	if err != nil {
+		log.Debug("send path failed: ", err)
+		return false, err
 	}
-	var hashRemote []byte
-	err = decoder.Decode(&hashRemote)
+	err = encoder.Encode(size)
 	if err != nil {
-		log.Fatal("Cient protocol remote hash error:", err)
-		return
+		log.Debug("send size failed: ", err)
+		return false, err
 	}
 
-	// Compare file hashes
-	hashLocal = fileHasher.Sum(nil)
-	if isHashDifferent(hashLocal, hashRemote) || FailPointFileHashMatch() {
-		log.Warn("hashLocal =", hashLocal)
-		log.Warn("hashRemote=", hashRemote)
-		err = &HashCollsisionError{}
-	} else {
-		retry = false // success, don't retry anymore
+	var ack bool
+	err = decoder.Decode(&ack)
+	if err != nil {
+		log.Debug("decode ack from server failed: ", err)
+		return false, err
 	}
+	log.Debugf("got SyncRequest ack back: %v", ack)
 
-	// Final retry negotiation
-	{
-		err1 := encoder.Encode(retry)
-		if err1 != nil {
-			log.Fatal("Cient protocol remote retry error:", err)
-		}
-		err1 = decoder.Decode(&statusRemote)
-		if err1 != nil {
-			log.Fatal("Cient protocol remote retry status error:", err)
-		}
-	}
-	return
+	return ack, nil
```

Comment:
What I meant is, check if `ack == false` here and if it is, return error.

Then the caller of function won't need to check both variables.